linkerd/app/core/src/errors.rs

```diff
@@ -44,11 +45,75 @@ pub enum Respond<B> {
     Http2 { is_grpc: bool },
 }
 
-impl<A, B: Default> respond::NewRespond<http::Request<A>> for NewRespond<B> {
-    type Response = http::Response<B>;
-    type Respond = Respond<B>;
+pub enum ResponseBody<B> {
+    NonGrpc(B),
+    Grpc {
+        inner: B,
+        trailers: Option<http::HeaderMap>,
+    },
+}
+
+impl<B: hyper::body::Payload> hyper::body::Payload for ResponseBody<B>
+where
+    B::Error: Into<Error>,
+{
+    type Data = B::Data;
+    type Error = B::Error;
 
-    fn new_respond(&self, req: &http::Request<A>) -> Self::Respond {
+    fn poll_data(&mut self) -> Poll<Option<Self::Data>, Self::Error> {
+        match self {
+            Self::NonGrpc(inner) => inner.poll_data(),
+            Self::Grpc { inner, trailers } => {
+                // should not be calling poll_data if we have set trailers derived from an error
+                assert!(trailers.is_none());
+                match inner.poll_data() {
+                    Err(error) => {
+                        let error = error.into();
+                        let mut error_trailers = http::HeaderMap::new();
+                        let code = set_grpc_status(&error, &mut error_trailers);
+                        warn!(
+                            "Handling body error: {:?} with grpc status {:?}",
+                            error, code
+                        );
```

Comment:
I think this should probably be at `debug` and not `warn`?  We mostly want to warn to surface proxy errors.

Also, we should use typical tracing metadata instead of formatting like:
```rust
debug!(%error, grpc.status = %code, "Handling gRPC stream failure");